pkg/resource/deploy/step_generator.go

```diff
@@ -179,6 +179,42 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, res
 		contract.Assert(len(steps) == 0)
 		return nil, res
 	}
+
+	// Check each proposed step against the relevant resource plan, if any
+	for _, s := range steps {
```

Comment:
This all looks reasonable - just stylistically I'm wondering if some of it could be pulled out into pure functions for clarity? I'm trying to calibrate my biases vs what's the preferred style here (or idiomatic Go..).. 

What seems to be going on here, there's a few side effects.

If `sg.deployment.plan != nil` then []Step is checked against it, with failures surfacing as result.Errorf. Could this be a pure func? Or is `resourcePlan.Ops = resourcePlan.Ops[1:]` also meaningfully mutates/ pops off Op? The order of the Ops and the order of the Steps must match then?

Then the other part is in ExperimentalPlans we  push the op `resourcePlan.Ops = append(resourcePlan.Ops, s.Op())`.

Somehow this is in the same for loop. I'm now wondering if that's really essential for this to be in the same for loop like the op queue appended to by L214 needs sometimes to be popped off in L191, like the side-effect interaction is essential?

Reply:
Also it seems that GenerateSteps has only *one* call site in the codebase. Does this logic better fit inside GenerateSteps because sequencing is really important, or could it be factored into "CheckPlans" or such and integrated where GenerateSteps call site is?

Reply:
> Then the other part is in ExperimentalPlans we push the op resourcePlan.Ops = append(resourcePlan.Ops, s.Op()).

Note that is the generation of *newPlans* vs the modification of the *currentPlan*. So yes all the mutation here is kinda needed.

> or could it be factored into "CheckPlans" or such and integrated where GenerateSteps call site is?

It would be really invasive to push it down into generateSteps where the step list is made.